src/main/java/ch/njol/skript/expressions/ExprColorOf.java

```diff
@@ -113,76 +172,119 @@ public Class<?>[] acceptChange(ChangeMode mode) {
 		if (FireworkEffect.class.isAssignableFrom(returnType))
 			return CollectionUtils.array(Color[].class);
 
+		switch (mode) { // items/blocks/entities don't accept these change modes
+			case ADD:
+			case REMOVE:
+			case REMOVE_ALL:
+				return null;
+		}
+
 		if (mode != ChangeMode.SET && !getExpr().isSingle())
 			return null;
 
-		if (Entity.class.isAssignableFrom(returnType))
-			return CollectionUtils.array(Color.class);
-		else if (Block.class.isAssignableFrom(returnType))
-			return CollectionUtils.array(Color.class);
-		if (ItemType.class.isAssignableFrom(returnType))
-			return CollectionUtils.array(Color.class);
-		return null;
+		return CollectionUtils.array(Color.class);
 	}
 
-	@SuppressWarnings("deprecated")
 	@Override
-	public void change(Event e, @Nullable Object[] delta, ChangeMode mode) {
-		if (delta == null)
-			return;
-		DyeColor color = ((Color) delta[0]).asDyeColor();
+	public void change(Event event, @Nullable Object[] delta, ChangeMode mode) {
+		DyeColor originalColor = (mode == ChangeMode.DELETE || mode == ChangeMode.RESET) ? null : ((Color) delta[0]).asDyeColor();;
 
-		for (Object o : getExpr().getArray(e)) {
-			if (o instanceof Item || o instanceof ItemType) {
-				ItemStack stack = o instanceof Item ? ((Item) o).getItemStack() : ((ItemType) o).getRandom();
+		for (Object obj : getExpr().getArray(event)) {
+			DyeColor color = originalColor; // reset
+			if (obj instanceof Item || obj instanceof ItemType || obj instanceof Slot) {
+				ItemStack stack;
+				if (obj instanceof ItemType) {
+					stack = ((ItemType) obj).getRandom();
+				} else if (obj instanceof Item) {
+					stack = ((Item) obj).getItemStack();
+				} else {
+					stack = ((Slot) obj).getItem();
+				}
 
 				if (stack == null)
 					continue;
 
-				MaterialData data = stack.getData();
-
-				if (!(data instanceof Colorable))
-					continue;
-
-				((Colorable) data).setColor(color);
-				stack.setData(data);
-
-				if (o instanceof Item)
-					((Item) o).setItemStack(stack);
-			} else if (o instanceof Block || o instanceof Colorable) {
-				Colorable colorable = getColorable(o);
+				if (!(obj instanceof Colorable)) {
```

Comment:
In what scenarios would something be both Colorable and an Item/ItemType/Slot? From my quick glance it seems only Blocks/Entities are Colorable.

Reply:
Good catch! the get method doesn't have that check as it's indeed wrong.